README.md
## Eyes.Storybook

Applitools Eyes SDK for Storybook

[![npm](https://img.shields.io/npm/v/@applitools/eyes.storybook.svg?style=for-the-badge)](https://www.npmjs.com/package/@applitools/eyes.storybook)

### Installation

Add your Applitools API key to your environment variables as `APPLITOOLS_API_KEY`

Install eyes.storybook as a local dev dependency in your tested project:

    npm install --save-dev @applitools/eyes.storybook

### Usage

Before running your project's build, you should run:

```
$ npx eyes-setup
```

The above command will update your .storybook/config.js so that the Eyes Storybook SDK can automatically grab stories to test.

You should then run your project's build (e.g, ```npm run build```)

Now you can run visual tests for all your stories using the following command:

```
$ npx eyes-storybook
```

### Advanced configuration

To change viewport sizes, application name, test name, etc., you can use a configuration file. 

Create configuration file called `applitools.config.js` in your project directory (the name can be changed using `--conf` CLI option, e.g. `npm run eyes-storybook -- --conf myconfig.js`, or add it to your package.json).

All available options are listed below:

```js
module.exports = {
    /* Server configuration */

    // `serverUrl` is the Eyes server URL that will be used during matching screenshots
    serverUrl: undefined, // default address stored in eyes.sdk.core

    // `proxy` defines the proxy server that will be used for requests to Applitools services
    // Should be a string in following format 'http://username:password@hostname:port/'
    proxy: undefined, // default is not set

    // `apiKey` is the Applitools API Key which can be found in the Admin Panel on website
    apiKey: process.env.APPLITOOLS_API_KEY,  // as default used value from environment variable


    /* App and test configuration */

    // `appName` is your application name that will be shown in test results
    appName: undefined, // as default used your package name from package.json

    // `viewportSize` is the required browser's viewport size or a list of sizes. It can be
    // an array of objects or a single object, e.g. {width: 800, height: 600}
    viewportSize: [
        {width: 800, height: 600}, // by default we use the value
    ],

    // `maxConcurrency` is a number of parallel browsers or connections to VisualGrid service
    maxConcurrency: 0, // default is 0, which means not limited connections to VisualGrid service
                       // or 10 for headless browsers (should be set 1 for non-headless browser)

    // `tapFilePath` is a path to TAP results file, the path is relative to directory
    // from which you run the test and should contains filename, e.g. './results.tap'
    tapFilePath: undefined, // by default we don't create the file anywhere


    /* Storybook configuration */

    // `storybookApp` is used to modify config according to your app. Usually, we don't need
    // the value, but you can force it by use one of values [react, vue, react-native, angular, polymer]
    storybookApp: undefined, // default is extracted from dependencies of your package.json 

    // `storybookVersion` is related to `storybookApp` and handled in the similar way, defines
    // which version of Storybook are you using, possible values [2, 3, 4]
    storybookVersion: undefined, // default is extracted from dependencies of your package.json

    // `storybookConfigDir` defines directory where to load Storybook configurations from.
    // The value will be passed to Storybook via `--config-dir` option
    storybookConfigDir: process.env.SBCONFIG_CONFIG_DIR || './.storybook', // Storybook default

    // `storybookStaticDir` defines directory where to load static files from, comma-separated list.
    // The value will be passed to Storybook via `--static-dir` option
    storybookStaticDir: process.env.SBCONFIG_STATIC_DIR, // Storybook default


    /* Only for VisualGrid mode */

    // `skipStorybookBuild` defines whether or not will be run `build-storybook` command.
    skipStorybookBuild: true, // if you don't use custom build command, we can do that for you,
                              // just set this to `false` and we will run `build-storybook` every time

    // `storybookOutputDir` defines directory where to store built files.
    // The value will be passed to Storybook via `--output-dir` option
    storybookOutputDir: process.env.SBCONFIG_OUTPUT_DIR || './storybook-static', // Storybook default


    /* Only for Selenium mode */
    
    // `useSelenium` defines a mode in which to work. VisualGrid mode creates a Storybook build
    // and send it to a VisualGrid service which creates a screenshots of each story (in a cloud).
    // Selenium mode starts a browsers locally and makes screenshots locally, after that send
    // only images for validation
    useSelenium: false, // default mode is VisualGrid (remote) mode,
                        // change this to `true` to use Selenium (local) mode

    // `storybookUrl` defines an address to an external Storybook server. Define this value
    // only in case if you don't want that starting Storybook was part of our process.
    // The value should be like 'http://localhost:9001/'
    storybookUrl: undefined, // by default we will start Storybook server in the process

    // `storybookPort` defines port on which we will start Storybook server. The value
    // is not related to `storybookUrl` and will be ignored if you specify both values
    storybookPort: process.env.SBCONFIG_PORT || 9001, // Storybook default

    // `storybookHost` defines host on which we will start Storybook server. The value
    // is similar to `storybookPort`, can't be used with `storybookUrl`
    storybookHost: process.env.SBCONFIG_HOSTNAME || 'localhost', // Storybook default

    // `seleniumUrl` defines address to selenium server. You can use the next url as 
    // an example: 'http://localhost:4444/wd/hub'
    seleniumUrl: undefined, // by default we start build-in selenium server

    // `capabilities` defines capabilities that will be passed to WebDriver when using local
    // testing or will be send as configuration of VisualGrid when using remote testing.
    // In Browser mode the `capabilities` directly passed to Selenium server, see docs
    // https://github.com/SeleniumHQ/selenium/wiki/DesiredCapabilities
    capabilities: { // by default we use chrome in headless mode
        platform: 'any', // local: the current system platform will be used
                         // remote: [] no currently supported values, will be added soon
        browserName: 'chrome', // local: make sure that you have required WebDriver in your PATH
                               // remote: ['chrome', 'firefox'], an array is also possible
        chromeOptions: { // used to set arguments for browser in local-only mode
                         // the name can be different depends of browser, see Selenium docs
            args: ['--headless', '--disable-gpu'], // any, e.g. '--force-device-scale-factor=2'
        },
    },


    /* Logging  */

    // `showLogs` defines whether or not you want to see logs. There are three possible values:
    // false - means no logs, only test results
    // true - some logs, about which story processing at the moment
    // 'verbose' - all available logs, report about each operation in the SDK
    showLogs: false, // default is disabled

    // `showEyesSdkLogs` defines whether or not you want to see logs from eyes.sdk.core.
    // Can be useful if you want to see information about connections to the services.
    // Same as with `showLogs`, there possible three values [false, true, 'verbose']
    showEyesSdkLogs: false, // default is disabled

    // `showStorybookOutput` defines whether or not you want to see Storybook output.
    // If Storybook server can't be started, or started with errors, set this option to true
    showStorybookOutput: false, // default is disabled
};
```

You can use content above as template for your `applitools.config.js`

### CLI Options

There is no required options, but some can be used to simplify working experience.

Below you can see output of the `--help` option.

```
Usage: eyes-storybook [options]

Options:
  --help            Show help                                                              [boolean]
  --version, -v     Show the version number                                                [boolean]
  --conf, -f        Path to Applitools configuration file [string] [default: "applitools.config.js"]
  --static-dir, -s  Directory where to load static files from, comma-separated list         [string]
  --output-dir, -o  Directory where the built result files are stored                       [string]
  --config-dir, -c  Directory where to load Storybook configurations from                   [string]
  --port, -p        Port to run Storybook                                                   [number]
  --host, -h        Host to run Storybook                                                   [string]
  --local, -l       Force to use Selenium mode                                             [boolean]
  --legacy          Use old implementation of VisualGrid test runner                       [boolean]
  --build           Enable building Storybook before testing                               [boolean]
  --exitcode, -e    If tests failed close with non-zero exit code                          [boolean]
  --info, -d        Display info about current running story                               [boolean]
  --verbose, --dd   Display data about current running method                              [boolean]
  --debug, --ddd    Display all possible logs and debug information                        [boolean]
```

### Independent Storybook Server

If you would like to run Storybook server out of the `eyes-storybook` execution, you should specify `storybookUrl` option in your `applitools.config.js` file and update Storybook's config file according to rules below.

### Automatic Build

If you would Eyes storybook to to automatically run `build-storybook` command when it is run, you can use the `--build` parameter to the `eyes-storybook` command.

### 'eyes-setup'

The ```eyes-setup``` command updates your .storybook/config.js file to include the following lines:

```
if (typeof window === 'object' && window.navigator && (/node\.js/i).test(window.navigator.userAgent)) {
  let addons = require('@storybook/addons').default;
  let Channel = require('@storybook/channels').default;
  addons.setChannel(new Channel({
    transport: {
      setHandler: function() {},
      send: function() {}
    }
  }));
}
```
at the beggining of the file and

```
if (typeof window === 'object') {
  window.__storybook_stories__ = require('@storybook/react').getStorybook();
}
```
at the end.

Inserting these lines manually to your config.js file is the same as using the ```eyes-setup``` command.

To access list of stories we need a way to access Storybook from browser. Please add lines below to Storybook's config file (default path is `.storybook/config.js`).

---


This guide is also available in the Applitools website:

- https://applitools.com/resources/tutorial/other/react_storybook



CHANGELOG.md
# Changelog

## [Unreleased](https://github.com/applitools/eyes.storybook/compare/v1.7.0...HEAD)
### Features
- Added output of logs from internal `build-storybook` execution
- Added prefix to logs from `build-storybook` or `start-storybook` when they display to general log channel
- Updated dependencies

## [1.7.0](https://github.com/applitools/eyes.storybook/compare/v1.6.1...v1.7.0) - 2018-06-11
### Features
- `chromedriver` dependency is not mandatory for Browser (local) mode, if you already have WebDriver in your PATH, you can skip this dependency
- In VisualGrid (remote) mode, you can set multiple browsers in `capabilities`.`browserName` property. Tests will be preformed on all of them
- Added `eyes-setup` script which allows automatically update Storybook's config (It adds code which exports Storybook to `window` object)
- Added new options `--config-dir`, `--static-dir`, `--output-dir`, `--host`, `--port`, to have similar options like Storybook has

### Bug Fixes
- Close WebDriver when all tests completed in Browser (local) mode

### BREAKING CHANGES
- Changed default value of `viewportSize` property. Now only one size (800x600) will be used
- Replaced next aliases from options `-h` is now belongs to `--host` instead of `--help`, `-c` belongs to `--config-dir` instead of `--conf`

## [1.6.1](https://github.com/applitools/eyes.storybook/compare/v1.6.0...v1.6.1) - 2018-06-05
### Bug Fixes
- Fixed en error which occurred on big bulk requests, because of empty objects in `/render-status` response
- Increased timeouts between `/render-status` requests

## [1.6.0](https://github.com/applitools/eyes.storybook/compare/v1.5.0...v1.6.0) - 2018-05-24
### BREAKING CHANGES
- Renamed `useVisualGrid` config property to `useSelenium` (inverted). Default mode not changed
- Renamed `storybookAddress` and `seleniumAddress` config properties to `storybookUrl` and `seleniumUrl` accordingly
- `Browser mode` is renamed to `Selenium mode`. Renamed Browser classes to Selenium, Rendering to VisualGrid
- `selenium-webdriver` and `chromedriver` modules were removed from dependencies. If you would like to use Selenium mode, you need to install them manually
- Use bulk `/render` and `/render-status` requests to VisualGrid, no concurrency limit (if for some reason you want to use old one, add `--legacy` option)

### Features
- Renamed several variables which contained reference to Storybook 3 (because they are used in Storybook 4 too)
- Progress bar updated, added display of number of processed stories
- Added ability to save results as TAP (Test Anything Protocol) file, you can do that by using `tapFilePath` config property
- Added ability to change platform and browser for VisualGrid testing. To do that use `capabilities`.`platform` and `capabilities`.`browserName` properties similar to Selenium config. Currently supported values are `chrome` and `firefox` for `browserName`, `platform` values will be added later
- Get default batch name from `APPLITOOLS_BATCH_NAME` env variable
- RenderRequest updated, now include render height too (previously only width was used)
- Updated dependencies

## [1.5.0](https://github.com/applitools/eyes.storybook/compare/v1.4.1...v1.5.0) - 2018-04-02
### BREAKING CHANGES
- Changed `--verbose` and `--debug` options' aliases to `-d` and `--dd` accordingly
- Renamed `useRenderer` config property to `useVisualGrid`. Changed default value to `true`
- Set `skipStorybookBuild` config property to `true` by default
- Added `--exitcode` option to use non-zero exit code when tests failed (previously this behaviour was active by default)
- Disabled printing logs by default

### Features
- Updated README file, now includes detailed descriptions
- Added CHANGELOG file
- Added log message which indicates when no config loaded
- Added `--build` option, to force storybook build 
- Added `--info` option, to display base logs
- Added `--local` option, to force use Browser mode
- If log output is disabled, added spinner which indicates processes

## [1.4.1](https://github.com/applitools/eyes.storybook/compare/v1.4.0...v1.4.1) - 2018-02-27
### Bug Fixes
- Increased request timeouts to wait while storybook/angular rebuild webpack
- Fixed extracting stories from builds where exists vendor.bundle.js and preview.bundle.js files 

### Features
- Added ability to set proxy server for external connections
- Added `--renderer` option which force to use render service

## [1.4.0](https://github.com/applitools/eyes.storybook/compare/v1.3.0...v1.4.0) - 2018-02-25
### Features
- User checkSingleWindow method, to reduce number of requests to eyes server

## [1.3.0](https://github.com/applitools/eyes.storybook/compare/v1.2.1...v1.3.0) - 2018-02-24
### Bug Fixes
- Added wait between get storybook bundle requests

### Features
- Added skipStorybookBuild config option
- Changed method to get stories from build to async

## [1.2.1](https://github.com/applitools/eyes.storybook/compare/v1.2.0...v1.2.1) - 2018-02-23
### BREAKING CHANGES
- Renamed `maxBrowsersCount` to `maxConcurrency`, and changed default

### Bug Fixes
- Fixed config validation

### Features
- Created DOM and resources before processing stories
- Removed url from RGridDom, use same DOM for each story

## [1.2.0](https://github.com/applitools/eyes.storybook/compare/v1.1.0...v1.2.0) - 2018-02-21
### Bug Fixes
- Fixed work with node 6
- Improved waiting for storybook server while it is starting

### Features
- Updated log messages
- Added support of React Native, Angular and Polymer
- License changed to Apache 2

## [1.1.0](https://github.com/applitools/eyes.storybook/compare/v1.0.2...v1.1.0) - 2018-02-12
### BREAKING CHANGES
- Package renamed to @applitools/eyes.storybook
- Exit code when tests failed changed to non-zero

### Bug Fixes
- Added reject promise if Storybook port already in use
- Improved work with storybook server

### Features
- Added `serverUrl` to configuration

lib/DefaultConfig.js
'use strict';

// See README for details
module.exports = {
  // Server
  serverUrl: undefined,
  proxy: undefined,
  apiKey: process.env.APPLITOOLS_API_KEY,

  // App & Test
  appName: process.env.APPLITOOLS_BATCH_NAME,
  viewportSize: [
    { width: 800, height: 600 },
  ],
  maxConcurrency: 0,
  tapFilePath: undefined,

  // Storybook
  storybookApp: undefined,
  storybookVersion: undefined,
  storybookConfigDir: process.env.SBCONFIG_CONFIG_DIR || './.storybook',
  storybookStaticDir: process.env.SBCONFIG_STATIC_DIR,

  // VisualGrid mode, default mode
  skipStorybookBuild: false,
  storybookOutputDir: process.env.SBCONFIG_OUTPUT_DIR || './storybook-static',

  // Selenium mode
  useSelenium: false,
  storybookUrl: undefined,
  storybookPort: process.env.SBCONFIG_PORT || 9001,
  storybookHost: process.env.SBCONFIG_HOSTNAME || 'localhost',
  seleniumUrl: undefined,
  capabilities: {
    platform: 'any',
    browserName: 'chrome',
    chromeOptions: {
      args: ['--headless', '--disable-gpu'],
    },
  },

  // Logs
  showLogs: false,
  showEyesSdkLogs: false,
  showStorybookOutput: false,
};
